packages/controllers/src/permissions/Permission.ts

```diff
@@ -0,0 +1,462 @@
+ * target name. In a way, the inverse of {@link ExtractPermissionTargetNames}.
+ *
+ * See the documentation for the distinction between target names and keys.
+ *
+ * @template Key - The target key type to extract from.
+ * @template Name - The name whose key to extract.
+ */
+export type ExtractPermissionTargetKey<
+  Key extends string,
+  Name extends string,
+> = Key extends Name ? Key : Extract<Key, KeyOfTargetName<Key, Name>>;
+
+/**
+ * Internal utility for extracting the members types of an array. The type
+ * evalutes to `never` if the specified type is the empty tuple or neither
+ * an array nor a tuple.
+ *
+ * @template ArrayType - The array type whose members to extract.
+ */
+type ExtractArrayMembers<ArrayType> = ArrayType extends
+  | readonly [...infer ArrayMembers]
+  | [...infer ArrayMembers]
+  ? ArrayMembers[number]
+  : never;
+
+/**
+ * A utility type for extracting the allowed caveat types for a particular
+ * permission from a permission specification type or type union.
+ *
+ * @template PermissionSpecification - The permission specification type to
+ * extract valid caveat types from.
+ * @template TargetName - The target name of the permission whose allowed
+ * caveats to extract.
+ */
+export type ExtractAllowedCaveatTypes<
+  PermissionSpecification extends PermissionSpecificationConstraint,
+  TargetName extends string,
+> = PermissionSpecification extends PermissionSpecificationConstraint & {
+  targetKey: ExtractPermissionTargetKey<
+    PermissionSpecification['targetKey'],
+    TargetName
+  >;
+}
+  ? ExtractArrayMembers<PermissionSpecification['allowedCaveats']>
+  : never;
+
+/**
+ * The options object of {@link constructPermission}.
+ *
+ * @template TargetPermission - The {@link Permission} that will be constructed.
+ */
+export type PermissionOptions<TargetPermission extends PermissionConstraint> = {
+  target: TargetPermission['parentCapability'];
+  /**
+   * The origin string of the subject that has the permission.
+   */
+  invoker: OriginString;
+
+  /**
+   * The GUID of the permission object.
+   * Assigned if not provided.
+   */
+  id?: string;
+
+  /**
+   * The caveats of the permission.
+   * See {@link Caveat}.
+   */
+  caveats?: NonEmptyArray<CaveatConstraint>;
+  // TODO: Refactor allowed caveats validation to enable this?
+  // caveats?: NonEmptyArray<ExtractCaveats<TargetPermission>>;
+};
+
+/**
+ * The default permission factory function. Naively constructs a permission from
+ * the inputs. Sets a default, random `id` if none is provided.
+ *
+ * @see {@link Permission} For more details.
+ *
+ * @template TargetPermission - The {@link Permission} that will be constructed.
+ * @param options - The options for the permission.
+ * @returns The new permission object.
+ */
+export function constructPermission<
+  TargetPermission extends PermissionConstraint,
+>(options: PermissionOptions<TargetPermission>): PermissionConstraint {
```

Comment:
What if we did this instead? 9299cae

Reply:
"This" being cast the returned object as `TargetPermission`. That seems to work, I removed all casts where `constructPermission` is called (the final one in 2f67293).

Reply:
I still don't understand what purpose this serves. Won't the type evaluate to `PermissionConstraint` at build time, since we don't know what the specific permission will be? If so this just adds complexity without adding any type safety.

Reply:
Not so! See e.g. here in `PermissionController.test.ts`: 
![image](https://user-images.githubusercontent.com/25517051/141196991-ce6ac449-93ee-4c12-8974-7117eb8eb0cd.png)

Reply:
That looks like a factory function. This thread was regarding the `constructPermission` function, used in the absence of a factory.

If there's a factory, we can know what the permission type will be statically because the factory is part of the spec, and each entry of the spec is for a specific permission. But for `constructPermission`, the specific permission type isn't known until runtime.

Reply:
Oh but I guess you're saying the type is still useful for the factory permission options. Fair enough.